commercetools-convenience/src/main/java/io/sphere/sdk/queries/QueryAllImpl.java

```diff
@@ -22,42 +24,140 @@ private QueryAllImpl(final QueryDsl<T, C> baseQuery, final long pageSize) {
+                                                         final Consumer<List<T>> consumer) {
+        final long totalPages = getTotalNumberOfPages(totalElements);
+        return LongStream.range(1, totalPages)
+                         .mapToObj(page -> queryPage(client, page)
+                             .thenApply(PagedQueryResult::getResults)
+                             .thenAccept(consumer)
+                             .toCompletableFuture())
+                         .collect(toList());
+    }
+
+    /**
+     * Given a total number of elements {@code totalElements} this method calculates the number of pages needed that
+     * would be needed to cover all the elements with set {@code pageSize} of {@link this} instance.
+     *
+     * @param totalElements number of elements to get the number of pages for.
+     * @return the total number of pages.
+     */
+    long getTotalNumberOfPages(final long totalElements) {
+        return (long) Math.ceil((double) totalElements / pageSize);
     }
 
+    /**
+     * Gets the results of {@link this} instance's query for a specific page with {@code pageNumber}.
+     *
+     * @param client     the CTP client that the query is run on.
+     * @param pageNumber the page number to get the results for.
+     * @return a future containing the results of the requested page of applying the query.
+     */
+    @Nonnull
     private CompletionStage<PagedQueryResult<T>> queryPage(final SphereClient client, final long pageNumber) {
         final QueryDsl<T, C> query = baseQuery
-                .withOffset(pageNumber * pageSize)
-                .withLimit(pageSize);
+            .withOffset(pageNumber * pageSize)
+            .withLimit(pageSize);
         return client.execute(query);
     }
 
-    @SuppressWarnings("rawtypes")
-    private CompletableFuture<List<T>> transformListOfFuturesToFutureOfLists(final List<CompletableFuture<List<T>>> futures) {
+    /**
+     * Given a list of futures, this method converts it to a future containing a list of results of these futures after
+     * executing them in parallel with {@link CompletableFuture#allOf(CompletableFuture[])}.
+     *
+     * @param futures list of futures.
+     * @param <S>     the type of the results of the futures.
+     * @return a future containing a list of the results of the input futures.
+     */
+    @Nonnull
+    private <S> CompletableFuture<List<S>> transformListOfFuturesToFutureOfLists(
+        final List<CompletableFuture<S>> futures) {
         final CompletableFuture[] futuresAsArray = futures.toArray(new CompletableFuture[futures.size()]);
         return CompletableFuture.allOf(futuresAsArray)
-                .thenApply(x -> futures.stream()
-                        .map(CompletableFuture::join)
-                        .flatMap(Collection::stream)
-                        .collect(Collectors.<T>toList()));
+                                .thenApply(x -> futures.stream()
```

Comment:
I don't fully understand this line. Why do we first convert the futures to an array and then then later use list of futures `x -> futures.stream()` ? What am I missing here? 

But looks like this code was there before you extended it 😉  But would be nice to understand it anyway 😄

Reply:
because of JDK function signature: `java.util.concurrent.CompletableFuture#allOf(CompletableFuture<?>... cfs)`  - it accepts an array, not a list. But later in `thenApply` it is nicer to work with List/Stream abstraction, instead of raw arrays.

Not sure if it is possible to simplify it here.

Reply:
@andrii-kovalenko-ct I see, it's because `CompletableFuture.allOf` returns `CompletionStage<Void>`